### PR TITLE
1785 free spec

### PR DESCRIPTION
**Pull request by mcodescu**
I have added the condition regarding free/cofree links. Should be fine now. I will remove the tracings after the code gets reviewed.

### CODE REVIEW

File: Static/AnalysisLibrary.hs [hunk 1]
```diff
@@ -269,8 +271,10 @@ anaLibDefn lgraph opts topLns libenv dg (Lib_defn ln alibItems pos ans) file
       ([], dg { globalAnnos = allAnnos }, libenv
       , lgraph, Map.empty) (map item alibItems)
   let dg1 = computeDGraphTheories libenv' $ markFree libenv'
-        $ markHiding libenv' $ fromMaybe dg' $ maybeResult
-        $ shortcutUnions dg'
+        $ markHiding libenv' 
+        $ fromMaybe dg' $ maybeResult
+        $ shortcutUnions 
+        dg'
```
Comment (tillmo) [inline Static/AnalysisLibrary.hs]: this is a mere reformatting. Is this needed?

File: Static/AnalysisLibrary.hs [hunk 2]
```diff
@@ -395,7 +400,7 @@ anaLibItem lg opts topLns currLn libenv dg eo itm =
               (alreadyDefined spstr) pos
       else
         -- let (_n, dg''') = addSpecNodeRT dg'' (UnitSig args body) $ show spn
-        return
+        return $
```
Comment (tillmo) [inline Static/AnalysisLibrary.hs]: the $ should be removed. Please return to the old code here.

File: Static/AnalysisStructured.hs [hunk 3]
```diff
@@ -125,11 +125,12 @@ insGTheory dg name orig (G_theory lid syn sig ind sens tind) =
...
-        (if tb then setThMapDG $ Map.insert (succ t) nth tMap else id) $
-        (if ind == startSigId
-         then setSigMapDG $ Map.insert (succ s) nsig sgMap else id)
-         $ insNodeDG (node, node_contents) dg)
+        (nsig_res,dg_res) = (NodeSig node nsig,
+          (if tb then setThMapDG $ Map.insert (succ t) nth tMap else id) $
+          (if ind == startSigId
+           then setSigMapDG $ Map.insert (succ s) nsig sgMap else id)
+           $ insNodeDG (node, node_contents) dg)
+    in (nsig_res,dg_res)
```
Comment (tillmo) [inline Static/AnalysisStructured.hs]: this is a mere reformulation. Is this needed?

File: Static/ComputeTheory.hs [hunk 4]
```diff
@@ -26,6 +26,7 @@ module Static.ComputeTheory
     , getImportNames
     ) where
 
+
```
Comment (tillmo) [inline Static/ComputeTheory.hs]: please drop this extra line

File: Static/ComputeTheory.hs [hunk 5]
```diff
@@ -102,19 +103,21 @@ computeLibEnvTheories le =
     in foldl upd Map.empty lns
 
 computeDGraphTheories :: LibEnv -> DGraph -> DGraph
-computeDGraphTheories le dgraph =
+computeDGraphTheories le dgraph =
```
Comment (tillmo) [inline Static/ComputeTheory.hs]: please drop this extra space

File: Static/ComputeTheory.hs [hunk 6]
```diff
@@ -102,19 +103,21 @@ computeLibEnvTheories le =
...
 
 computeDGraphTheoriesAux :: LibEnv -> DGraph -> DGraph
-computeDGraphTheoriesAux le dgraph =
+computeDGraphTheoriesAux le dgraph = let
+  tnodes = topsortedNodes dgraph 
+ in 
   foldl (\ dg l@(n, lbl) -> updatePending dg lbl
     (n, recomputeNodeLabel le dg l))
-     dgraph $ topsortedNodes dgraph
+     dgraph tnodes
```
Comment (tillmo) [inline Static/ComputeTheory.hs]: this is a mere reformulation. Is this needed?

File: Static/ComputeTheory.hs [hunk 7]
```diff
@@ -102,19 +103,21 @@ computeLibEnvTheories le =
...
 
 recomputeNodeLabel :: LibEnv -> DGraph -> LNode DGNodeLab -> DGNodeLab
-recomputeNodeLabel le dg l@(n, lbl) =
+recomputeNodeLabel le dg l@(n, lbl) =
```
Comment (tillmo) [inline Static/ComputeTheory.hs]: please remove the extra space

File: Static/AnalysisLibrary.hs (hunk 1 above)
Comment (mcodescu) [reply Static/AnalysisLibrary.hs]: It's not, it was caused by me trying to find out how much I can comment. I have reverted it.

File: Static/AnalysisLibrary.hs (hunk 2 above)
Comment (mcodescu) [reply Static/AnalysisLibrary.hs]: done.

File: Static/AnalysisStructured.hs (hunk 3 above)
Comment (mcodescu) [reply Static/AnalysisStructured.hs]: This was needed for tracing. I have reverted it.

File: Static/ComputeTheory.hs (hunk 4 above)
Comment (mcodescu) [reply Static/ComputeTheory.hs]: Done

File: Static/ComputeTheory.hs (hunk 5 above)
Comment (mcodescu) [reply Static/ComputeTheory.hs]: Done

File: Static/ComputeTheory.hs (hunk 6 above)
Comment (mcodescu) [reply Static/ComputeTheory.hs]: No (unless you prefer to have less $ and more names in use). I have added it for tracing. Reverted.

File: Static/ComputeTheory.hs (hunk 7 above)
Comment (mcodescu) [reply Static/ComputeTheory.hs]: Done

**Comment by tillmo** [comment]
good, many thanks. However, if you look at https://github.com/spechub/Hets/pull/1798/files , you will see that a few places, there are still changes which are just concerning white space.

**Comment by tillmo** [comment]
good. Now you must rebase to master:
```
git checkout 1785_free_spec
git fetch
git rebase origin/master
git push -f
```

**Comment by mcodescu** [comment]
I have tried "Update branch", as I wanted to see how that goes. Looks like it worked. Merging now.

**Pull request merged**